### PR TITLE
Improve and clarify the logic when speculatively comparing old/new code to ensure semantics are the same.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
I'm not a fan of the existing style, which i feel was a little too brittle/unclear.  The new style attempts to be much crisper about what it is doing and why.

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -214,16 +214,18 @@ private static void AnalyzeArrayInitializer(SyntaxNodeAnalysisContext context)
...
+        // values)`) will not get simplification.  We can revisit this in the future to see if that warrants a more
+        // expensive check that involves checking the consitutuent elements of the literal.
         var speculationAnalyzer = new SpeculationAnalyzer(
             topmostExpression,
-            s_nullLiteralExpression,
+            s_emptyCollectionExpression,
             semanticModel,
             cancellationToken,
-            skipVerificationForReplacedNode: true,
+            skipVerificationForReplacedNode: false,
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:228]: changed this (was a hack originally).  we were previously making a change, but then skipping checking the actual node we were replacing.  now we check it like all the rest above it to make sure it's ok.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs [hunk 2]
```diff
@@ -722,28 +722,65 @@ private bool ReplacementBreaksQueryClause(QueryClauseSyntax originalClause, Quer
                 !SymbolInfosAreCompatible(originalClauseInfo.OperationInfo, newClauseInfo.OperationInfo);
         }
 
-        protected override bool ReplacementIntroducesErrorType(ExpressionSyntax originalExpression, ExpressionSyntax newExpression)
+        protected override bool ReplacementIntroducesDisallowedNullType(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs:725]: broke this into two checks.  a check for error-types (always bad), and a check for a null type (sometimes bad, sometimes ok).

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs [hunk 3]
```diff
@@ -581,11 +595,22 @@ private bool ReplacementChangesSemanticsForNode(SyntaxNode currentOriginalNode,
...
+                Debug.Assert(this.SemanticRootOfOriginalExpression.DescendantNodesAndSelf().Contains(originalExpression));
+                RoslynDebug.AssertNotNull(newExpression);
+                Debug.Assert(this.SemanticRootOfReplacedExpression.DescendantNodesAndSelf().Contains(newExpression));
+
+                var originalTypeInfo = this.OriginalSemanticModel.GetTypeInfo(originalExpression);
+                var newTypeInfo = this.SpeculativeSemanticModel.GetTypeInfo(newExpression);
+
+                // If we didn't have an error before, but now we got one, that's bad and should block conversion in all cases.
+                if (newTypeInfo.Type.IsErrorType() && !originalTypeInfo.Type.IsErrorType())
+                    return true;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs:610]: this is inlining the error check from before.  it is *not* something that should be overridable.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs [hunk 4]
```diff
@@ -581,11 +595,22 @@ private bool ReplacementChangesSemanticsForNode(SyntaxNode currentOriginalNode,
...
+
+                if (ReplacementIntroducesDisallowedNullType(originalExpression, newExpression, originalTypeInfo, newTypeInfo))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs:612]: it is then followed by the null-type check.  where the language *can* override if null is a problem or not.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -214,16 +214,17 @@ private static void AnalyzeArrayInitializer(SyntaxNodeAnalysisContext context)
...
+        // Note: this does mean certain unambiguous cases with overloads (like `Goo(int[] values)` vs `Goo(string[]
+        // values)`) will not get simplification.  We can revisit this in the future to see if that warrants a more
+        // expensive check that involves checking the consitutuent elements of the literal.
         var speculationAnalyzer = new SpeculationAnalyzer(
             topmostExpression,
-            s_nullLiteralExpression,
+            s_emptyCollectionExpression,
             semanticModel,
             cancellationToken,
-            skipVerificationForReplacedNode: true,
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:226]: changed this (was a hack originally). we were previously making a change, but then skipping checking the actual node we were replacing. now we check it like all the rest above it to make sure it's ok.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs [hunk 6]
```diff
@@ -81,14 +75,12 @@ protected override SyntaxNode GetSemanticRootForSpeculation(ExpressionSyntax exp
...
-                node.Kind() == SyntaxKind.ThisConstructorInitializer ||
-                node.Kind() == SyntaxKind.BaseConstructorInitializer ||
-                node.Kind() == SyntaxKind.EqualsValueClause ||
-                node.Kind() == SyntaxKind.ArrowExpressionClause;
+            return node is StatementSyntax(kind: not SyntaxKind.Block) or TypeSyntax or CrefSyntax ||
+                node.Kind() is SyntaxKind.Attribute or
+                               SyntaxKind.ThisConstructorInitializer or
+                               SyntaxKind.BaseConstructorInitializer or
+                               SyntaxKind.EqualsValueClause or
+                               SyntaxKind.ArrowExpressionClause;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs:83]: just a simplification.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs [hunk 7]
```diff
@@ -506,7 +498,7 @@ protected override bool ReplacementChangesSemanticsForNodeLanguageSpecific(Synta
             }
             else if (currentOriginalNode.Kind() == SyntaxKind.ImplicitArrayCreationExpression)
             {
-                return !TypesAreCompatible((ImplicitArrayCreationExpressionSyntax)currentOriginalNode, (ImplicitArrayCreationExpressionSyntax)currentReplacedNode);
+                return !TypesAreCompatible((ExpressionSyntax)currentOriginalNode, (ExpressionSyntax)currentReplacedNode);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/SpeculationAnalyzer.cs:501]: incorrect logic before. the point of this analyzer is that the old construct and new construct might not match, but semantics shoudl still be preserved.  so, in this case, the new construct just needs to be an expression, not an implicit array like before.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/SpeculationAnalyzer.vb [hunk 8]
```diff
@@ -42,10 +42,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Utilities
         ''' True if semantic analysis should fail when any of the invocation expression ancestors of <paramref name="expression"/> in original code has overload resolution failures.
         ''' </param>        
         Public Sub New(expression As ExpressionSyntax, newExpression As ExpressionSyntax, semanticModel As SemanticModel, cancellationToken As CancellationToken, Optional skipVerificationForReplacedNode As Boolean = False, Optional failOnOverloadResolutionFailuresInOriginalCode As Boolean = False)
-            MyBase.New(expression, newExpression, semanticModel, cancellationToken, skipVerificationForReplacedNode, failOnOverloadResolutionFailuresInOriginalCode)
+            MyBase.New(expression, newExpression, semanticModel, skipVerificationForReplacedNode, failOnOverloadResolutionFailuresInOriginalCode, cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/SpeculationAnalyzer.vb:45]: VB still has skipVerificationForReplacedNode in a few places.  I don't love it and i'd like to remove in teh future as well.  but out of scope for this PR.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs [hunk 9]
```diff
@@ -581,11 +595,22 @@ private bool ReplacementChangesSemanticsForNode(SyntaxNode currentOriginalNode,
             else if (currentOriginalNode is TExpressionSyntax originalExpression)
             {
                 var newExpression = (TExpressionSyntax)currentReplacedNode;
-                if (!ImplicitConversionsAreCompatible(originalExpression, newExpression) ||
-                    ReplacementIntroducesErrorType(originalExpression, newExpression))
-                {
+                if (!ImplicitConversionsAreCompatible(originalExpression, newExpression))
+                    return true;
```
Comment (genlu) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs:599]: missing newline

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs [hunk 10]
```diff
@@ -581,11 +595,22 @@ private bool ReplacementChangesSemanticsForNode(SyntaxNode currentOriginalNode,
...
+                RoslynDebug.AssertNotNull(originalExpression);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs:600]:
```suggestion

                RoslynDebug.AssertNotNull(originalExpression);
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs:599]: added.  thanks.

**Review by genlu — APPROVED** [review]
:shipit:

**Review by akhera99 — APPROVED** [review]

**Pull request merged**